player_alsa.go

```diff
@@ -0,0 +1,121 @@
+// +build linux
+// +build !js
+// +build !android
+// +build !ios
+
+package oto
+
+// #cgo linux LDFLAGS: -lasound
+// #include <alsa/asoundlib.h>
+// #include "player_alsa.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"unsafe"
+)
+
+type player struct {
+	handle     *C.snd_pcm_t
+	buf        []byte
+	bufSamples int
+}
+
+func alsaError(err C.int) error {
+	return errors.New(C.GoString(C.snd_strerror(err)))
+}
+
+func newPlayer(sampleRate, numChans, bytesPerSample, bufferSizeInBytes int) (*player, error) {
+	var p player
+
+	// open a default ALSA audio device for blocking stream playback
+	errCode := C.snd_pcm_open(&p.handle, C.CString("default"), C.SND_PCM_STREAM_PLAYBACK, 0)
+	if errCode < 0 {
+		return nil, alsaError(errCode)
+	}
+
+	var (
+		// sample format, either SND_PCM_FORMAT_S8 or SND_PCM_FORMAT_S16_LE
+		format C.snd_pcm_format_t
+		// bufferSize is the total size of the main circular buffer fullness of this buffer
+		// oscilates somewhere between bufferSize and bufferSize-periodSize
+		bufferSize = C.snd_pcm_uframes_t(bufferSizeInBytes / (numChans * bytesPerSample))
```

Comment:
I prefer using `:=` as much as possible. In this case, `bufferSize := ...` is better to me. (I'm fine to leave this as it is if you want, but I might change this in the future).